pymc_extras/inference/deterministic_advi/dadvi.py

```diff
@@ -0,0 +1,228 @@
+    Accurate, and Even More Black Box. Journal of Machine Learning
+    Research, 25(18), 1–39.
+    """
+
+    model = pymc.modelcontext(model) if model is None else model
+
+    initial_point_dict = model.initial_point()
+    n_params = DictToArrayBijection.map(initial_point_dict).data.shape[0]
+
+    var_params, objective = create_dadvi_graph(
+        model,
+        n_fixed_draws=n_fixed_draws,
+        random_seed=random_seed,
+        n_params=n_params,
+    )
+
+    f_fused, f_hessp = _compile_functions_for_scipy_optimize(
+        objective,
+        [var_params],
+        compute_grad=True,
+        compute_hessp=True,
+        compute_hess=False,
+    )
+
+    result = minimize(
+        f_fused, np.zeros(2 * n_params), method="trust-ncg", jac=True, hessp=f_hessp
+    )
+
+    opt_var_params = result.x
+    opt_means, opt_log_sds = np.split(opt_var_params, 2)
+
+    # Make the draws:
+    draws_raw = np.random.randn(n_draws, n_params)
+    draws = opt_means + draws_raw * np.exp(opt_log_sds)
+    draws_arviz = unstack_laplace_draws(draws, model, chains=1, draws=n_draws)
+
+    transformed_draws = transform_draws(
+        draws_arviz, model, keep_untransformed=keep_untransformed
+    )
+
+    return transformed_draws
+
+
+def create_dadvi_graph(
+    model: Model,
+    n_params: int,
+    n_fixed_draws: int = 30,
+    random_seed: int = 2,
+) -> Tuple[TensorVariable, TensorVariable]:
+    """
+    Sets up the DADVI graph in pytensor and returns it.
+
+    Parameters
+    ----------
+    model : pm.Model
+        The PyMC model to be fit.
+
+    n_params: int
+        The total number of parameters in the model.
+
+    n_fixed_draws : int
+        The number of fixed draws to use.
+
+    random_seed: int
+        The random seed to use for the fixed draws.
+
+    Returns
+    -------
+    Tuple[TensorVariable, TensorVariable]
+        A tuple whose first element contains the variational parameters,
+        and whose second contains the DADVI objective.
+    """
+
+    # Make the fixed draws
+    state = np.random.RandomState(random_seed)
+    draws = state.randn(n_fixed_draws, n_params)
+
+    inputs = model.continuous_value_vars + model.discrete_value_vars
+    initial_point_dict = model.initial_point()
+    logp = model.logp()
+
+    # Graph in terms of a flat input
+    [logp], flat_input = join_nonshared_inputs(
+        point=initial_point_dict, outputs=[logp], inputs=inputs
+    )
+
+    var_params = pt.vector(name="eta", shape=(2 * n_params,))
+
+    means = var_params[:n_params]
+    log_sds = var_params[n_params:]
+
+    draw = pt.vector(name="draw", shape=(n_params,))
+    sample = means + pt.exp(log_sds) * draw
+
+    # Graph in terms of a single sample
+    logp_draw = pytensor.clone_replace(logp, replace={flat_input: sample})
+    draw_matrix = pt.constant(draws)
+
+    # Vectorise
+    logp_vectorized_draws = pytensor.graph.vectorize_graph(
+        logp_draw, replace={draw: draw_matrix}
+    )
```

Comment:
I don't think you need the intermediate `clone_replace` step?

```suggestion
    draw_matrix = pt.constant(draws)
    samples = means + pt.exp(log_sds) * draw_matrix
    
    logp_vectorized_draws = pytensor.graph.vectorize_graph(
        logp, replace={flat_input: samples}
    )
```